tests/search_users_test.php

```diff
@@ -0,0 +1,130 @@
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Search users
+ *
+ * @package    tool_mergeusers
+ * @copyright  2024 Leon Stringer <leon.stringer@ntlworld.com>
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace tool_mergeusers;
+
+defined('MOODLE_INTERNAL') || die();
+
+require_once(__DIR__ . '/../lib/mergeusersearch.php');
+
+/**
+ * Tests for searching for users.
+ */
+final class search_users_test extends \advanced_testcase {
+    /**
+     * Test search criteria works with PostgreSQL because this will generate an
+     * error when comparing and integer column (mdl_user.id) with a string.
+     * @dataProvider search_criteria
+     */
+    public function test_pgsqlsearch($searchfield, $input, $count): void {
+        global $DB;
+
+        // Skip tests if not using PostgreSQL.
+        if ($DB->get_dbfamily() != 'postgres') {
+            $this->markTestSkipped('PostgreSQL-only test');
+        }
+
+        $mus = new \MergeUserSearch();
+        $this->assertIsArray($mus->search_users($input, $searchfield));
+    }
+
+    /**
+     * Test deleted users are not returned with any search criteria.
+     * @dataProvider search_criteria
+     */
+    public function test_nodeletedusers($searchfield, $input, $count): void {
+        $this->resetAfterTest(true);
+
+        $deleteduser = $this->getDataGenerator()->create_user([
+            'username' => 'student1', 'email' => 'student1@example.com',
+            'firstname' => 'Student', 'lastname' => 'One',
+            'idnumber' => 'ID001',
+        ]);
+        delete_user($deleteduser);
+        $this->getDataGenerator()->create_user([
+            'username' => 'student1', 'email' => 'student1@example.com',
+            'firstname' => 'Student', 'lastname' => 'One',
+            'idnumber' => 'ID001',
+        ]);
+
+        if ($searchfield === 'id') {
+            $input = $deleteduser->id;
+        } else if ($searchfield === 'email') {
+            $input = md5($deleteduser->username);
+        }
+
+        $mus = new \MergeUserSearch();
+        $this->assertEquals($count,
+                    count($mus->search_users($input, $searchfield)));
+    }
+
+    /**
+     * Test various allowed values for MergeUserSearch->search_users()'s
+     * $searchfield parameter.
+     */
+    public static function search_criteria(): array {
+        return [
+            'id' => [
+                'searchfield' => 'id',
+                'input' => '',
+                'count' => 0,
+            ],
+            'id2' => [
```

Comment:
I can see that you removed the test with null as input value. Even though it would be an edge case (not possible using Moodle administration to merge users), it can be possible if someone integrates it programatically and undetectedly pass some null value over there.  Please, add again that test case with null as input value. Also, removing this test case, your patch to distinct between id IS NULL and so, does not apply now. Actually, it would apply to empty string (since it is not a number), but is a false positive.

Reply:
I've increased the data in the data provider, it includes `'input' => null`.

Reply:
thanks!